### PR TITLE
Rework main level logic

### PR DESCRIPTION
**Pull request by ShovelD**

### CODE REVIEW

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/IAbility.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+namespace Enemies.Bosses.Abilities
+{
+    [HideReferenceObjectPicker]
+    public interface IAbility
+    {
+        public void Initialize(Boss boss, TankImpl tank);
+        public void ExecuteAbility();
+        public void EnableAbility();
+        public void DisableAbility();
+        public bool IsActive { get; }
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/IAbility.cs:14]: Add default implementation logic to Enable|disable ability directly in interface

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Boss.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        [OdinSerialize]
+        [LabelText("Abilities")]
+        private List<IAbility> abilities = new();
+
+        [OdinSerialize]
+        private TankImpl tank;
+
+        public event Action OnDeath;
+
+        public string EnemyName => throw new System.NotImplementedException();
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Boss.cs:24]: Why it's not implemented?

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Boss.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+        public void Awake()
+        {
+            foreach (var ability in abilities)
+            {
+                ability.Initialize(this, tank);
+            }
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Boss.cs:31]: Check formatting settings, var should not be used

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Boss.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        public void Initialize(TankImpl tank)
+        {
+            this.tank = tank;
+        }
+
+        public void TakeDamage(float damageAmount)
+        {
+            OnDeath?.Invoke();
+        }
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Boss.cs:42]: What The Focuses in this code, where correct implementation?

File: Tank Survivors/Assets/Scripts/Enemies/EnemyProducers/BossProducer.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+using UnityEngine;
+
+namespace Enemies.EnemyProducers
+{
+    [Serializable]
+    public class BossProducer
+    {
+        [AssetsOnly]
+        [SerializeField]
+        [AssetList(CustomFilterMethod = "EnemiesFilter")]
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/Enemies/EnemyProducers/BossProducer.cs:13]: use nameof() in attributes

File: Tank Survivors/Assets/Scripts/General/EnemyGenerator.cs [hunk 6]
```diff
@@ -53,5 +66,36 @@ private void Update()
                 toRemove.Clear();
             }
         }
+
+        private void TryGenerateBoss()
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/General/EnemyGenerator.cs:70]: Try means return true or false, and why you trying to generate it? Just create it

File: Tank Survivors/Assets/Scripts/General/EnemyGenerator.cs [hunk 7]
```diff
@@ -53,5 +66,36 @@ private void Update()
...
+
+        private void RemoveEnemiesFromScene()
+        {
+            var enemies = enemyRoot.GetComponentsInChildren<Transform>().ToList();
+            _ = enemies.Remove(enemyRoot);
+            foreach (var enemy in enemies)
+            {
+                Destroy(enemy.gameObject);
+            }
+        }
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/General/EnemyGenerator.cs:99]: Don't do this, just stopping producing enemies will be enough.

**Review by LeshaTola — CHANGES_REQUESTED** [review]
I think you can now leave this implementation with ability and take next task that will rewre all enemy structure to modular system, and bosses will be the same as enemies but with unique modules. System will bw the same as system in weapons modules.

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/MovementAbility.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using Sirenix.Serialization;
+using Tank;
+using UnityEngine;
+
+namespace Enemies.Bosses.Abilities
+{
+    [Serializable]
+    public class MovementAbility : IAbility
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/MovementAbility.cs:9]:
```suggestion
    public class Movement : IAbility
```

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/IAbility.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        public void Initialize(Boss boss, TankImpl tank);
+        public void ExecuteAbility();
+        public void EnableAbility()
+        {
+            IsActive = true;
+        }
+        public void DisableAbility()
+        {
+            IsActive = false;
+        }
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/IAbility.cs:19]:
```suggestion
        public void Execute();
        public void Enable()
        {
            IsActive = true;
        }
        public void Disable()
        {
            IsActive = false;
        }
```

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/MovementAbility.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            if (IsActive)
+            {
+                Vector3 direction = CalculateDirection();
+                boss.transform.position += direction * speed * Time.deltaTime;
+                RotatateToTank(direction);
+            }
+        }
+
+        public Vector3 CalculateDirection()
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/MovementAbility.cs:36]:
```suggestion
        private Vector3 CalculateDirection()
```

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/MovementAbility.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            return (tank.transform.position - boss.transform.position).normalized;
+        }
+
+        public void RotatateToTank(Vector3 direction)
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Abilities/MovementAbility.cs:41]:
```suggestion
        private void RotatateToTank(Vector3 direction)
```

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/BossStats.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using Sirenix.Serialization;
+
+namespace Enemies.Bosses
+{
+    [Serializable]
+    public class BossStats
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/BossStats.cs:7]:
```suggestion
    public class Stats
```

**Review by Kiuh — APPROVED** [review]

File: Tank Survivors/Assets/Scripts/Enemies/Bosses/Stats.cs [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+namespace Enemies.Bosses
+{
+    [Serializable]
+    public class Stats
+    {
+        [OdinSerialize]
+        public float Health { get; private set; }
+    }
+}
```
Comment (LeshaTola) [inline Tank Survivors/Assets/Scripts/Enemies/Bosses/Stats.cs:12]: maybe you should do it like enemies. Using SO and structs

**Review by LeshaTola — APPROVED** [review]

**Pull request merged**